File: tareas/3/ChaconHugo-ValdezSebastian/compara_planif.cpp
#include <bits/stdc++.h>
using namespace std;

struct Proceso {
    char nombre;
    int llegada;
    int duracion;
    int restante;
    int finalizacion;
};

// ---------------- UTILIDADES ----------------

void imprimir_resultados(string nombre, vector<Proceso> p, string timeline) {
    double T=0, E=0, P=0;

    for (auto &proc : p) {
        int t = proc.finalizacion - proc.llegada;
        int e = t - proc.duracion;
        double p_i = (double)t / proc.duracion;

        T += t;
        E += e;
        P += p_i;
    }

    int n = p.size();
    cout << "  " << nombre << ": T=" << T/n
         << ", E=" << E/n
         << ", P=" << P/n << "\n";

    cout << "  " << timeline << "\n";
}

// ---------------- FCFS ----------------

void FCFS(vector<Proceso> procesos) {
    sort(procesos.begin(), procesos.end(),
         [](auto &a, auto &b){ return a.llegada < b.llegada; });

    int tiempo = 0;
    string timeline;

    for (auto &p : procesos) {
        while (tiempo < p.llegada) {
            timeline += "-";
            tiempo++;
        }

        for (int i = 0; i < p.duracion; i++) {
            timeline += p.nombre;
            tiempo++;
        }

        p.finalizacion = tiempo;
    }

    imprimir_resultados("FCFS", procesos, timeline);
}

// ---------------- ROUND ROBIN ----------------

void RR(vector<Proceso> procesos, int q, string nombre) {
    queue<int> cola;
    int n = procesos.size();
    vector<bool> en_cola(n, false);

    int tiempo = 0, completados = 0;
    string timeline;

    while (completados < n) {

        for (int i = 0; i < n; i++) {
            if (!en_cola[i] && procesos[i].llegada <= tiempo) {
                cola.push(i);
                en_cola[i] = true;
            }
        }

        if (cola.empty()) {
            timeline += "-";
            tiempo++;
            continue;
        }

        int i = cola.front(); cola.pop();

        int uso = min(q, procesos[i].restante);

        for (int t = 0; t < uso; t++) {
            timeline += procesos[i].nombre;
            tiempo++;

            for (int j = 0; j < n; j++) {
                if (!en_cola[j] && procesos[j].llegada <= tiempo) {
                    cola.push(j);
                    en_cola[j] = true;
                }
            }
        }

        procesos[i].restante -= uso;

        if (procesos[i].restante == 0) {
            procesos[i].finalizacion = tiempo;
            completados++;
        } else {
            cola.push(i);
        }
    }

    imprimir_resultados(nombre, procesos, timeline);
}

// ---------------- SPN ----------------

void SPN(vector<Proceso> procesos) {
    int n = procesos.size();
    vector<bool> terminado(n, false);
    int tiempo = 0, completados = 0;
    string timeline;

    while (completados < n) {

        int idx = -1;
        int min_t = INT_MAX;

        for (int i = 0; i < n; i++) {
            if (!terminado[i] && procesos[i].llegada <= tiempo &&
                procesos[i].duracion < min_t) {
                min_t = procesos[i].duracion;
                idx = i;
            }
        }

        if (idx == -1) {
            timeline += "-";
            tiempo++;
            continue;
        }

        for (int i = 0; i < procesos[idx].duracion; i++) {
            timeline += procesos[idx].nombre;
            tiempo++;
        }

        procesos[idx].finalizacion = tiempo;
        terminado[idx] = true;
        completados++;
    }

    imprimir_resultados("SPN", procesos, timeline);
}

// ---------------- FEEDBACK (FB) ----------------

void FB(vector<Proceso> procesos) {
    int n = procesos.size();
    vector<queue<int>> colas(3);

    vector<bool> en_cola(n, false);

    int tiempo = 0, completados = 0;
    string timeline;

    while (completados < n) {

        for (int i = 0; i < n; i++) {
            if (!en_cola[i] && procesos[i].llegada <= tiempo) {
                colas[0].push(i);
                en_cola[i] = true;
            }
        }

        int nivel = -1;
        for (int i = 0; i < 3; i++) {
            if (!colas[i].empty()) {
                nivel = i;
                break;
            }
        }

        if (nivel == -1) {
            timeline += "-";
            tiempo++;
            continue;
        }

        int i = colas[nivel].front();
        colas[nivel].pop();

        int q = pow(2, nivel); // 1,2,4

        int uso = min(q, procesos[i].restante);

        for (int t = 0; t < uso; t++) {
            timeline += procesos[i].nombre;
            tiempo++;

            for (int j = 0; j < n; j++) {
                if (!en_cola[j] && procesos[j].llegada <= tiempo) {
                    colas[0].push(j);
                    en_cola[j] = true;
                }
            }
        }

        procesos[i].restante -= uso;

        if (procesos[i].restante == 0) {
            procesos[i].finalizacion = tiempo;
            completados++;
        } else {
            if (nivel < 2)
                colas[nivel+1].push(i);
            else
                colas[nivel].push(i);
        }
    }

    imprimir_resultados("FB", procesos, timeline);
}

// ---------------- MAIN ----------------

int main() {
    srand(time(0));

    for (int ronda = 1; ronda <= 5; ronda++) {
        cout << "- Ronda " << ronda << ":\n";

        vector<Proceso> procesos;
        int n = 5;

        for (int i = 0; i < n; i++) {
            Proceso p;
            p.nombre = 'A' + i;
            p.llegada = rand() % 10;
            p.duracion = rand() % 5 + 1;
            p.restante = p.duracion;
            procesos.push_back(p);
        }

        sort(procesos.begin(), procesos.end(),
             [](auto &a, auto &b){ return a.llegada < b.llegada; });

        for (auto &p : procesos) {
            cout << p.nombre << ": " << p.llegada
                 << ", t=" << p.duracion << "; ";
        }
        cout << "\n";

        FCFS(procesos);

        for (auto &p : procesos) p.restante = p.duracion;
        RR(procesos, 1, "RR1");

        for (auto &p : procesos) p.restante = p.duracion;
        RR(procesos, 4, "RR4");

        for (auto &p : procesos) p.restante = p.duracion;
        SPN(procesos);

        for (auto &p : procesos) p.restante = p.duracion;
        FB(procesos);

        cout << "\n";
    }

    return 0;
}

File: tareas/3/CruzLizbeth/ComparaPlanif.py
# Tarea 3
# El siguiente programa tiene como objetivo comparar los diferentes mecanismos de planificadores, van desde los mas sencillo FCFS
# RR(q=1 y q=4), SPN y FB (multinivel)
# Autora: Cruz Manriquez Lizbeth
# ************************************************************
# Importación de biblioteca
import random

# Para la simulación de los procesos, se define una función que los genere:

# Generar procesos

def generar_procesos():
    procesos = []
    for i in range(5):
        nombre = chr(65 + i)
        llegada = random.randint(0, 5)
        duracion = random.randint(1, 6)
        procesos.append([nombre, llegada, duracion])
    
    procesos.sort(key=lambda x: x[1])
    return procesos


# Para realizar la comparación entre los diferentes mecanismos pplanificadores
# se definen las métricas:

# Calcular métricas

def metricas(procesos, fin):
    T = 0 #Tiempo total desde que llega el proceso
    E = 0 #Tiempo de espera, donde el proceso permacene en cola
    P = 0 #Penalización, la relación entre tiempo total y tiempo de ejecución
    
    for p in procesos:
        nombre, llegada, duracion = p
        t_final = fin[nombre]
        
        t = t_final - llegada
        e = t - duracion
        p_ind = t / duracion
        
        T += t
        E += e
        P += p_ind
    
    n = len(procesos)
    return round(T/n,2), round(E/n,2), round(P/n,2)


# Primer mecanismo de planificación: First Come First Served
# Función FCFS

def fcfs(procesos):
    tiempo = 0
    linea = ""
    fin = {}

    for p in procesos:
        nombre, llegada, duracion = p
        
        if tiempo < llegada:
            tiempo = llegada
        
        for i in range(duracion):
            linea += nombre
            tiempo += 1
        
        fin[nombre] = tiempo
    
    return linea, metricas(procesos, fin)


# Segundo mecanismo, Round Robin (q=1 y q=4)
# Función Round Robin

def rr(procesos, q):
    tiempo = 0
    cola = []
    linea = ""
    fin = {}

    restantes = {p[0]: p[2] for p in procesos}

    i = 0
    while True:
        while i < len(procesos) and procesos[i][1] <= tiempo:
            cola.append(procesos[i][0])
            i += 1
        
        if len(cola) == 0:
            if i >= len(procesos):
                break
            tiempo += 1
            continue
        
        actual = cola.pop(0)
        ejec = min(q, restantes[actual])

        for j in range(ejec):
            linea += actual
            tiempo += 1

            while i < len(procesos) and procesos[i][1] <= tiempo:
                cola.append(procesos[i][0])
                i += 1
        
        restantes[actual] -= ejec

        if restantes[actual] > 0:
            cola.append(actual)
        else:
            fin[actual] = tiempo

    return linea, metricas(procesos, fin)


# Tercer mecanismo, Shortest Process Next
# Función SPN

def spn(procesos):
    tiempo = 0
    lista = []
    pendientes = procesos.copy()
    linea = ""
    fin = {}

    while len(pendientes) > 0 or len(lista) > 0:
        
        while len(pendientes) > 0 and pendientes[0][1] <= tiempo:
            lista.append(pendientes.pop(0))
        
        if len(lista) == 0:
            tiempo += 1
            continue
        
        lista.sort(key=lambda x: x[2])
        actual = lista.pop(0)

        nombre, llegada, duracion = actual

        for i in range(duracion):
            linea += nombre
            tiempo += 1
        
        fin[nombre] = tiempo

    return linea, metricas(procesos, fin)


# Se realiza un cuarto mecanismo Feedback Multinivel
# FB (multinivel simple)

def fb(procesos):
    tiempo = 0
    colas = [[], [], []]
    quantums = [1, 2, 4]

    linea = ""
    fin = {}
    restantes = {p[0]: p[2] for p in procesos}

    i = 0
    while True:
        while i < len(procesos) and procesos[i][1] <= tiempo:
            colas[0].append(procesos[i][0])
            i += 1
        
        nivel = -1
        for j in range(3):
            if len(colas[j]) > 0:
                nivel = j
                break
        
        if nivel == -1:
            if i >= len(procesos):
                break
            tiempo += 1
            continue
        
        actual = colas[nivel].pop(0)
        ejec = min(quantums[nivel], restantes[actual])

        for k in range(ejec):
            linea += actual
            tiempo += 1

            while i < len(procesos) and procesos[i][1] <= tiempo:
                colas[0].append(procesos[i][0])
                i += 1
        
        restantes[actual] -= ejec

        if restantes[actual] > 0:
            if nivel < 2:
                colas[nivel+1].append(actual)
            else:
                colas[nivel].append(actual)
        else:
            fin[actual] = tiempo

    return linea, metricas(procesos, fin)


# Función principal
# MAIN

for ronda in range(5):
    print("\n- Ronda", ronda+1)

    procesos = generar_procesos()

    # imprimir procesos y total
    total = 0
    linea_proc = ""
    for p in procesos:
        linea_proc += f"{p[0]}: {p[1]}, t={p[2]}; "
        total += p[2]

    print(" ", linea_proc.strip(), f"(tot:{total})")

    # FCFS
    l, m = fcfs(procesos)
    print(f"  FCFS: T={m[0]}, E={m[1]}, P={m[2]}")
    print("  ", l)

    # RR q=1
    l, m = rr(procesos, 1)
    print(f"  RR1: T={m[0]}, E={m[1]}, P={m[2]}")
    print("  ", l)

    # RR q=4
    l, m = rr(procesos, 4)
    print(f"  RR4: T={m[0]}, E={m[1]}, P={m[2]}")
    print("  ", l)

    # SPN
    l, m = spn(procesos)
    print(f"  SPN: T={m[0]}, E={m[1]}, P={m[2]}")
    print("  ", l)

    # FB
    l, m = fb(procesos)
    print(f"  FB: T={m[0]}, E={m[1]}, P={m[2]}")
    print("  ", l)
